std/string.d

```diff
@@ -5798,36 +5798,57 @@ unittest
  * function, or any of the conversion functions.
  *
  * Params:
- *     s = the string to check
+ *     s = the string or random access range to check
  *     bAllowSep = accept separator characters or not
  *
  * Returns:
  *     $(D bool)
  */
-bool isNumeric(S)(S s, bool bAllowSep = false) if (isSomeString!S)
+bool isNumeric(S)(S s, bool bAllowSep = false) if (isSomeString!S ||
+    (isRandomAccessRange!S &&
+    hasSlicing!S &&
+    isSomeChar!(ElementEncodingType!S) &&
+    !isInfinite!S))
 {
     import std.algorithm.comparison : among;
+    import std.ascii : isASCII;
+    import std.utf : byChar;
 
-    if (s.empty)
+    // auto-decoding special case, we're only comparing to characters
+    // in the ASCII range so there's no reason to decode
+    static if (isSomeString!S)
+    {
+        import std.utf : byCodeUnit;
+        auto codeUnits = s.byCodeUnit;
+    }
+    else
+    {
+        alias codeUnits = s;
+    }
+
+    if (codeUnits.empty)
         return false;
 
     // Check for NaN (Not a Number) and for Infinity
-    if (s.among!((a, b) => icmp(a, b) == 0)
+    if (codeUnits.among!((a, b) => icmp(a, b.byChar) == 0)
```

Comment:
Need to `.save` here, or the range is consumed. Test with a reference type, e.g. std.range.refRange:

```d
dstring z = "0";
isNumeric(refRange(&z)); // triggers assert in std.range.primitives
```

Also, this is `std.uni.icmp`, right? Could do with an ASCII one, no?